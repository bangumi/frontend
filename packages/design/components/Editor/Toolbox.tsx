import React, { FC } from 'react'
import { Bold, Image, Italic, Link, Size, Underscore } from '@bangumi/icons'

const Toolbox: FC = () => {
  return (
    <div className="bgm-editor__header">
      <Bold />
      <Italic />
      <Underscore />
      <Image />
      <Link />
      <Size />
    </div>
  )
}

export default Toolbox
