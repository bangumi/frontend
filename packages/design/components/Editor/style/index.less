@import '../../../theme/base';

@toolbox-with: 215px;

.bgm-editor {
  &__container {
    display: inline-flex;
    flex-direction: column;
    min-height: 162px;
    border: 2px solid @gray-10;
    border-radius: 12px;
    padding: 10px 0 0 12px;
  }

  &__header {
    display: flex;
    justify-content: space-between;
    align-items: center;
    height: 26px;
    width: @toolbox-with;
    margin-bottom: 10px;

    > svg {
      cursor: pointer;
    }
  }

  &__text {
    flex: 1 0 auto;
    padding: 0;
    min-width: @toolbox-with;
    width: 682px;
    border: none;
    outline: none;
    color: @gray-100;
    line-height: 26px;

    &::placeholder {
      color: @gray-60;
    }
  }
}
