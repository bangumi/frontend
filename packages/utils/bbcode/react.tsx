import React from 'react';

import { convert } from './convert';
import { Parser } from './parser';
import type { CodeNodeTypes, NodeTypes, VNode } from './types';

/* eslint-disable no-useless-escape */
const toCamelCase = (s: string) => s.replace(/(\-\w)/g, (k) => k[1]?.toUpperCase() ?? '');

const joinClassName = (cls: string | string[]) => {
  if (Array.isArray(cls)) {
    return cls.join(' ');
  }
  return cls;
};

const convertToReactProps = (node: VNode, key?: React.Key) => {
  const { props, style, className } = node;

  const transformedStyle: Record<string, string> = style
    ? Object.entries(style).reduce((pre, [key, val]) => {
        return {
          ...pre,
          [toCamelCase(key)]: val,
        };
      }, {})
    : {};

  const transformedClassName = className !== undefined ? joinClassName(className) : undefined;

  const reactProps = {
    ...props,
    style: transformedStyle,
    className: transformedClassName,
    key,
  };

  return reactProps;
};

export const renderNode = (node: NodeTypes, key?: React.Key): React.ReactElement | string => {
  if (typeof node === 'string') {
    const splittedStr: Array<string | React.ReactElement> = node.split(/\n/g);

    if (splittedStr.length < 2) {
      return splittedStr[0] ?? '';
    }
    const children = splittedStr.reduce<Array<string | React.ReactElement>>((pre, curr, idx) => {
      if (idx > 0) {
        return [...pre, React.createElement('br', null), curr];
      }
      return [...pre, curr];
    }, []);
    return React.createElement(React.Fragment, { key }, ...children);
  }

  const { type, children } = node;

  if (children) {
    return React.createElement(
      type,
      convertToReactProps(node, key),
      ...children.map((child, idx) => renderNode(child, idx)),
    );
  }
  return React.createElement(type, convertToReactProps(node, key));
};

export const render = (rawStr: string): Array<React.ReactElement | string> => {
  const nodes: CodeNodeTypes[] = new Parser(rawStr).parse();

  return nodes.map((node, idx) => renderNode(convert(node, {}), idx));
};
