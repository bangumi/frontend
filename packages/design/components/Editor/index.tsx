/* eslint-disable react/prop-types */
import React, { FC, ChangeEvent, useRef } from 'react'
import Toolbox from './Toolbox'

export interface EditorProps {
  onChange: (e: ChangeEvent<HTMLTextAreaElement>) => void
}

const setInputValue = (el: HTMLTextAreaElement, content: string, shouldUpdateSelection = true): void => {
  const pos = el.value.length + content.lastIndexOf('[')
  el.value += content
  shouldUpdateSelection && el.setSelectionRange(pos, pos)
  el.focus()
}

const Editor: FC<EditorProps> = ({
  onChange
}) => {
  const ref = useRef<HTMLTextAreaElement>(null)
  const handleToolboxClick = (type: string, payload?: any): void => {
    const el = ref.current
    if (!el) {
      return
    }
    switch (type) {
      case 'bold': {
        setInputValue(el, '[b][/b]')
        break
      }
      case 'italic': {
        setInputValue(el, '[i][/i]')
        break
      }
      case 'underscore': {
        setInputValue(el, '[u][/u]')
        break
      }
      case 'image': {
        const value = prompt('请输入图片链接')
        if (value === null) break
        setInputValue(el, `[img]${value}[/img]`, false)
        break
      }
      case 'link': {
        const value = prompt('请输入链接地址')
        if (value === null) break
        setInputValue(el, `[url=${value}]连接描述[/url]`)
        break
      }
      case 'size': {
        const value = prompt('请输入字体大小')
        if (value === null) break
        setInputValue(el, `[size=${value}][/size]`)
        break
      }
    }
  }
  return (
    <div className="bgm-editor__container">
      <Toolbox handleClickEvent={handleToolboxClick} />
      <textarea className="bgm-editor__text" placeholder="想聊点什么的呢..." onChange={onChange} ref={ref} />
    </div>
  )
}

export default Editor
