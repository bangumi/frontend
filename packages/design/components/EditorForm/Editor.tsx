import { size } from 'lodash';
import type { FocusEventHandler } from 'react';
import React, {
  forwardRef,
  useCallback,
  useEffect,
  useImperativeHandle,
  useRef,
  useState,
} from 'react';

import Toolbox from './Toolbox';

export interface EditorProps {
  placeholder?: string;
  /** 是否显示工具栏 */
  showToolbox?: boolean;
  /** 是否显示字数统计 */
  showWordCount?: boolean;
  /** textarea 通过键盘按下提交触发事件 */
  onConfirm?: (value: string) => void;
  /**
   * @default empty string
   */
  value?: string;
  /** 内容改变时的回调函数 */
  onChange?: (value: string) => void;
  /** 在被渲染时是否自动获取焦点 */
  autoFocus?: boolean;
  /** 文本域初始行数 */
  rows?: number;
  name?: string;
  onBlur?: FocusEventHandler<HTMLTextAreaElement>;
}

const keyToEvent: Record<string, string> = {
  b: 'bold',
  i: 'italic',
  u: 'underscore',
  l: 'link',
  s: 'size',
  p: 'image',
} as const;

const Editor = forwardRef<HTMLTextAreaElement, EditorProps>(
  (
    {
      placeholder,
      showToolbox = true,
      showWordCount = true,
      onConfirm,
      value = '',
      onChange,
      autoFocus,
      rows,
      name,
      onBlur,
    },
    ref,
  ) => {
    const [selection, setSelection] = useState<[number, number]>();
    const innerRef = useRef<HTMLTextAreaElement>(null);

    useImperativeHandle<HTMLTextAreaElement | null, HTMLTextAreaElement | null>(
      ref,
      () => innerRef.current,
    );

    useEffect(() => {
      // 用来确保输入光标在自动插入的quote内容之后。
      const el = innerRef.current;
      el?.setSelectionRange(el.value.length, el.value.length);
    }, []);

    // content改变时应重新设置选择范围
    useEffect(() => {
      if (!selection) return;
      const [start, end] = selection;
      const el = innerRef.current;
      el?.setSelectionRange(start, end);
      el?.focus();
    }, [selection]);

    const updateContent = useCallback(
      (newContent: string): void => {
        setSelection(undefined);
        onChange?.(newContent);
      },
      [setSelection, onChange],
    );

    const insertTag = useCallback(
      (el: HTMLTextAreaElement, prefix: string, suffix: string, selected = ''): void => {
        const preStart = el.selectionStart;
        const preEnd = el.selectionEnd;
        const preValue = el.value;
        updateContent(
          `${preValue.slice(0, preStart)}${prefix}${selected}${suffix}${preValue.slice(preEnd)}`,
        );
        // 未选择内容，插入BBCode标签后应将光标置于标签中间
        if (selected.length === 0) {
          const cursorPos = preStart + prefix.length + selected.length;
          setSelection([cursorPos, cursorPos]);
        }
        el.focus();
      },
      [updateContent, setSelection],
    );

    // 使用useCallback避免在内容改变时触发Toolbox的渲染
    const handleToolboxEvent = useCallback(
      (type: string, payload?: unknown): void => {
        const el = innerRef.current;
        if (!el) {
          return;
        }
        const preStart = el.selectionStart;
        const preEnd = el.selectionEnd;
        const selected = el.value.slice(preStart, preEnd);
        switch (type) {
          case 'bold': {
            insertTag(el, '[b]', '[/b]', selected);
            break;
          }
          case 'italic': {
            insertTag(el, '[i]', '[/i]', selected);
            break;
          }
          case 'underscore': {
            insertTag(el, '[u]', '[/u]', selected);
            break;
          }
          case 'image': {
            const value = prompt('请输入图片链接');
            if (value === null) break;
            insertTag(el, '[img]', '[/img]', value);
            break;
          }
          case 'link': {
            const value = prompt('请输入链接地址');
            if (value === null) break;
            const prefix = `[url=${value}]`;
            const suffix = '[/url]';
            insertTag(el, prefix, suffix, selected || '链接描述');
            if (!selected) {
              // 选中“链接描述”
              const start = preEnd + prefix.length;
              const end = start + 4;
              setSelection([start, end]);
            }
            break;
          }
          case 'size': {
            const value = prompt('请输入字体大小');
            if (value === null) break;
            insertTag(el, `[size=${value}]`, '[/size]', selected);
            break;
          }
        }
      },
      [innerRef, insertTag, setSelection],
    );

    const handleKeyDown = useCallback(
      (e: React.KeyboardEvent<HTMLTextAreaElement>) => {
        // Ctrl + Enter & Alt + S 触发提交
        if (
          ((e.ctrlKey || e.metaKey) && e.key === 'Enter') ||
          (e.altKey && e.key.toLowerCase() === 's')
        ) {
          onConfirm?.(innerRef.current!.value);
          e.preventDefault();
          return;
        }
        // https://bgm.tv/help/bbcode
        if (e.ctrlKey || e.metaKey) {
          const key = e.key.toLowerCase();
          const event = keyToEvent[key];
          if (event) {
            handleToolboxEvent(event);
            e.preventDefault();
          }
        }
      },
      [onConfirm, handleToolboxEvent],
    );

    return (
      <div className='bgm-editor__container'>
        <div className='bgm-editor__toolbar'>
          {showToolbox && <Toolbox handleClickEvent={handleToolboxEvent} />}
          {showWordCount && value && (
            <div className='bgm-editor__wordcount'>已输入 {size(value)} 字</div>
          )}
        </div>
        <textarea
          className='bgm-editor__text'
          placeholder={placeholder}
          ref={innerRef}
          value={value}
          autoFocus={autoFocus}
          onChange={(e) => {
            updateContent(e.target.value);
          }}
          onKeyDown={handleKeyDown}
          rows={rows}
          name={name}
          onBlur={onBlur}
        />
      </div>
    );
  },
);

export default Editor;
