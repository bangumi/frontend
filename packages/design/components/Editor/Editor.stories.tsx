import React from 'react'
import { ComponentStory, ComponentMeta } from '@storybook/react'
import Editor from '.'

const componentMeta: ComponentMeta<typeof Editor> = {
  title: 'Modern/Editor',
  component: Editor
}

export default componentMeta

const Template: ComponentStory<typeof Editor> = (args) => {
  return <Editor {...args} />
}

export const Usage = Template.bind({})
