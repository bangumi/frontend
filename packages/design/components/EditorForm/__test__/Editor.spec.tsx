import { fireEvent, render, screen } from '@testing-library/react';
import React, { useState } from 'react';

import type { EditorProps } from '../Editor';
import Editor from '../Editor';

const TestEditor = (props: EditorProps) => {
  const [value, setValue] = useState('');
  return <Editor value={value} onChange={setValue} {...props} />;
};

const initTextareaTest = (
  props: EditorProps,
): { textarea: HTMLTextAreaElement; setValue: (value: string) => void } => {
  props.placeholder = props.placeholder ?? 'placeholder';
  const { getByPlaceholderText } = render(<TestEditor {...props} />);
  const textarea = getByPlaceholderText(props.placeholder) as HTMLTextAreaElement;
  const setValue = (value: string) => fireEvent.change(textarea, { target: { value } });
  return { textarea, setValue };
};

const keyToEvent = {
  b: 'bold',
  i: 'italic',
  u: 'underscore',
  l: 'link',
  s: 'size',
  p: 'image',
};

function doSelectionTest(
  textarea: HTMLTextAreaElement,
  type: string,
  mockValue: string,
  selected = false,
): void {
  if (selected) {
    switch (type) {
      case 'bold': {
        expect(textarea).toHaveTextContent('Hello [b]World[/b]');
        expect(textarea.value).toBe('Hello [b]World[/b]');
        expect(textarea.selectionStart).toBe(18);
        expect(textarea.selectionEnd).toBe(18);
        break;
      }
      case 'italic': {
        expect(textarea.value).toBe('Hello [i]World[/i]');
        expect(textarea.selectionStart).toBe(18);
        expect(textarea.selectionEnd).toBe(18);
        break;
      }
      case 'underscore': {
        expect(textarea.value).toBe('Hello [u]World[/u]');
        expect(textarea.selectionStart).toBe(18);
        expect(textarea.selectionEnd).toBe(18);
        break;
      }
      case 'image': {
        expect(textarea.value).toBe(`Hello [img]${mockValue}[/img]`);
        expect(textarea.selectionStart).toBe(71);
        expect(textarea.selectionEnd).toBe(71);
        break;
      }
      case 'link': {
        expect(textarea.value).toBe(`Hello [url=${mockValue}]World[/url]`);
        expect(textarea.selectionStart).toBe(77);
        expect(textarea.selectionEnd).toBe(77);
        break;
      }
      case 'size': {
        expect(textarea.value).toBe(`Hello [size=${mockValue}]World[/size]`);
        expect(textarea.selectionStart).toBe(79);
        expect(textarea.selectionEnd).toBe(79);
      }
    }
  } else {
    switch (type) {
      case 'bold': {
        expect(textarea.value).toBe('[b][/b]');
        expect(textarea.selectionStart).toBe(3);
        expect(textarea.selectionEnd).toBe(3);
        break;
      }
      case 'italic': {
        expect(textarea.value).toBe('[i][/i]');
        expect(textarea.selectionStart).toBe(3);
        expect(textarea.selectionEnd).toBe(3);
        break;
      }
      case 'underscore': {
        expect(textarea.value).toBe('[u][/u]');
        expect(textarea.selectionStart).toBe(3);
        expect(textarea.selectionEnd).toBe(3);
        break;
      }
      case 'image': {
        expect(textarea.value).toBe(`[img]${mockValue}[/img]`);
        expect(textarea.selectionStart).toBe(65);
        expect(textarea.selectionEnd).toBe(65);
        break;
      }
      case 'link': {
        expect(textarea.value).toBe(`[url=${mockValue}]链接描述[/url]`);
        expect(textarea.selectionStart).toBe(60);
        expect(textarea.selectionEnd).toBe(64);
        break;
      }
      case 'size': {
        expect(textarea.value).toBe(`[size=${mockValue}][/size]`);
        expect(textarea.selectionStart).toBe(61);
        expect(textarea.selectionEnd).toBe(61);
      }
    }
  }
}

describe('EditorForm > Editor', () => {
  it('render correctly', () => {
    const { asFragment } = render(<Editor />);
    expect(asFragment()).toMatchSnapshot();
  });

  it('ref should forward to textarea tag', () => {
    const ref = React.createRef<HTMLTextAreaElement>();
    render(<Editor ref={ref} />);
    expect(ref.current?.tagName).toBe('TEXTAREA');
  });

  it('placeholder props', () => {
    const { getByPlaceholderText } = render(<Editor placeholder='Hello' />);
    expect(getByPlaceholderText('Hello')).toBeInTheDocument();
  });

  it('showToolbox props', () => {
    const { container } = render(<Editor showToolbox={false} />);
    expect(container.querySelector('.bgm-editor__toolbox')).not.toBeInTheDocument();
  });

  it('showWordCount props', () => {
    const { container } = render(<Editor showWordCount={false} />);
    expect(container.querySelector('.bgm-editor__wordcount')).not.toBeInTheDocument();
  });

  it('click toolbox should have correct behavior', () => {
    const { textarea, setValue } = initTextareaTest({ placeholder: 'Hello' });

    const mockValue = 'https://lain.bgm.tv/pic/cover/l/65/19/364450_9lB1T.jpg';
    const prompt = vi.spyOn(window, 'prompt').mockImplementation(() => mockValue);

    ['bold', 'italic', 'underscore', 'image', 'link', 'size'].forEach((type) => {
      // init
      setValue('');
      const el = screen.getByTestId(type);
      fireEvent.click(el);
      doSelectionTest(textarea, type, mockValue);

      if (type === 'image' || type === 'link' || type === 'size') {
        expect(prompt).toHaveBeenCalled();
      } else {
        expect(prompt).not.toHaveBeenCalled();
      }
      prompt.mockClear();

      // Click toolbox -> textarea should be focused
      expect(textarea).toHaveFocus();
    });
  });

  it('click toolbox with selection', () => {
    const { textarea, setValue } = initTextareaTest({ placeholder: 'Hello' });

    const mockValue = 'https://lain.bgm.tv/pic/cover/l/65/19/364450_9lB1T.jpg';
    const prompt = vi.spyOn(window, 'prompt').mockImplementation(() => mockValue);

    ['bold', 'italic', 'underscore', 'image', 'link', 'size'].forEach((type) => {
      // init
      setValue('Hello World');
      textarea.selectionStart = 6;
      textarea.selectionEnd = 11;

      const el = screen.getByTestId(type);
      fireEvent.click(el);
      doSelectionTest(textarea, type, mockValue, true);

      if (type === 'image' || type === 'link' || type === 'size') {
        expect(prompt).toHaveBeenCalled();
      } else {
        expect(prompt).not.toHaveBeenCalled();
      }
      prompt.mockClear();

      // Click toolbox -> textarea should be focused
      expect(textarea).toHaveFocus();
    });
  });

  it('onConfirm keyboard event', () => {
    const onConfirm = vi.fn();
    const { textarea } = initTextareaTest({ placeholder: 'Hello', onConfirm });

    textarea.value = 'test111';
    fireEvent.keyDown(textarea, { key: 'Enter', ctrlKey: true });
    expect(onConfirm).lastCalledWith('test111');

    textarea.value = 'test2';
    fireEvent.keyDown(textarea, { key: 's', altKey: true });
    expect(onConfirm).lastCalledWith('test2');

    textarea.value = 'test3';
    fireEvent.keyDown(textarea, { key: 'S', altKey: true });
    expect(onConfirm).lastCalledWith('test3');
  });

  it('BBCode editor keyboard event', () => {
    const { textarea, setValue } = initTextareaTest({ placeholder: 'Hello' });

    const mockValue = 'https://lain.bgm.tv/pic/cover/l/65/19/364450_9lB1T.jpg';
    const prompt = vi.spyOn(window, 'prompt').mockImplementation(() => mockValue);

    for (const key of Object.keys(keyToEvent)) {
      const type = keyToEvent[key as keyof typeof keyToEvent];
      // init
      setValue('');

      fireEvent.keyDown(textarea, { key, ctrlKey: true });
      doSelectionTest(textarea, type, mockValue);

      // Uppercase
      setValue('');
      fireEvent.keyDown(textarea, { key: key.toUpperCase(), ctrlKey: true });
      doSelectionTest(textarea, type, mockValue);

      if (type === 'image' || type === 'link' || type === 'size') {
        expect(prompt).toHaveBeenCalled();
      } else {
        expect(prompt).not.toHaveBeenCalled();
      }
      prompt.mockClear();
    }
  });

  it('BBCode editor keyboard event with selection', () => {
    const { textarea, setValue } = initTextareaTest({ placeholder: 'Hello' });

    const mockValue = 'https://lain.bgm.tv/pic/cover/l/65/19/364450_9lB1T.jpg';
    const prompt = vi.spyOn(window, 'prompt').mockImplementation(() => mockValue);

    for (const key of Object.keys(keyToEvent)) {
      const type = keyToEvent[key as keyof typeof keyToEvent];
      // init
      setValue('Hello World');
      textarea.selectionStart = 6;
      textarea.selectionEnd = 11;

      fireEvent.keyDown(textarea, { key, ctrlKey: true });
      doSelectionTest(textarea, type, mockValue, true);

      // Uppercase
      setValue('Hello World');
      textarea.selectionStart = 6;
      textarea.selectionEnd = 11;

      fireEvent.keyDown(textarea, { key: key.toUpperCase(), ctrlKey: true });
      doSelectionTest(textarea, type, mockValue, true);

      if (type === 'image' || type === 'link' || type === 'size') {
        expect(prompt).toHaveBeenCalled();
      } else {
        expect(prompt).not.toHaveBeenCalled();
      }
      prompt.mockClear();
    }
  });

  it('word count is working when input contains unicode', () => {
    const { getByText } = render(<Editor value='123👍' />);
    expect(getByText('已输入 4 字')).toBeInTheDocument();
  });
});
