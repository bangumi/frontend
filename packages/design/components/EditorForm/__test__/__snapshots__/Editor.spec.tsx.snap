// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`EditorForm > Editor > render correctly 1`] = `
<DocumentFragment>
  <div
    class="bgm-editor__container"
  >
    <div
      class="bgm-editor__toolbar"
    >
      <div
        class="bgm-editor__toolbox"
      >
        <div
          data-testid="bold"
          title="粗体 [Ctrl+B]"
        />
        <div
          data-testid="italic"
          title="斜体 [Ctrl+I]"
        />
        <div
          data-testid="underscore"
          title="下划线 [Ctrl+U]"
        />
        <div
          data-testid="image"
          title="图片 [Ctrl+P]"
        />
        <div
          data-testid="link"
          title="链接 [Ctrl+L]"
        />
        <div
          data-testid="size"
          title="文字尺寸 [Ctrl+S]"
        />
      </div>
    </div>
    <textarea
      class="bgm-editor__text"
    />
  </div>
</DocumentFragment>
`;
